### ISSUE TITLE
Linter suggestion: detect vector/list appending in loops

### ISSUE THREAD

**Issue by maelle** [jimhester/lintr#2871]
From https://github.com/ropensci-review-tools/goodpractice/issues/8

cc @Bisaloo

**Comment by maelle** [comment]
In https://github.com/r-lib/lintr/issues/48 there was

>  Iterative appending to vectors within loops rather than pre-allocation the vectors and filling. (the former is quadratic in time complexity, as R has to keep re-sizing the vector on every assignment.

```r
# bad
for (i in seq_len(10)) {
  x <- c(x, i)
}

# good
x <- integer(10)
for (i in seq_len(10)) {
  x[i] <- i
}
```

**Comment by MichaelChirico** [comment]
we may have a closed issue around this... the problem is detecting statically that the vector size can 100% be determined ex-ante, i.e. avoiding false positives.

it's a really important linter, though, so maybe we should approach this incrementally -- start from some really simple and obvious, narrow examples and slowly extend the rule to include more.
